### PR TITLE
fix(instrumentation-mongoose): Fix instrumentation for Mongoose Model methods, save() and remove(), by passing options argument to original method calls

### LINKED ISSUE #1678: Transaction is not used in Mongoose create operation after OpenTelemetry auto-instrumentation is added
### What version of OpenTelemetry are you using?
    "@opentelemetry/api": "^1.5.0",
    "@opentelemetry/auto-instrumentations-node": "^0.39.2",

### What version of Node are you using?
v16.15.1

### What did you do?
When running some database operation inside the transaction
```
const User = db.model('User', new Schema({ name: String }));
const session = await User.startSession();
await session.startTransaction();
const [newStaff] = await User.create([ {name: 'staffA'} ], {session});
await session.abortTransaction();
await session.endSession();

```

### What did you expect to see?
It is expected the transaction will be aborted, and user with name "staffA" will not be created in the database.

### What did you see instead?
The transaction was not rolled back, and the record was created in the database.

### Additional Info
I am using "mongoose": "^6.2.1"
The transaction works with `findOneAndUpdate`, for example the user name is not changed to `staffB` in the following case
```
const User = db.model('User', new Schema({ name: String }));
const [newStaff] = await User.create([ {name: 'staffA'} ]);
const session = await User.startSession();
await session.startTransaction();
await User.findOneAndUpdate({name: 'staffA'}, {name: 'staffB'}).session(session);
await session.abortTransaction();
await session.endSession();
```
However, when session is used in the `create` operation, the `lastUse` in `serverSession` is not updated across the whole function, which means the session is not used at all.

### PR DESCRIPTION
**Pull request by ferrelucas**
## Which problem is this PR solving?
Fixes #1678 

This PR fixes the issue where transactions cannot be used for the Model methods `save()` and `remove()` (and their aliases eg.  `create()`) because after instrumentation is applied to these methods, specifically when both `options` and `callback` are defined, the final method call loses the first argument `option` making the final call not take the `option.session` into consideration when making database operations.
This could lead to save updates in transactions not being properly rolled back as well as multiple operations that are supposed to share the same session not being able to do so because their calls would eventually be made without the `session` reference.

## Short description of the changes

- Pass the original `args` into `handleCallbackResponse()`
- Updates the implementation of `handleCallbackResponse()` so that we create `newArgs` including the original `options` and the instrumented`callback` function.
- Added tests to cover the base case of empty-object but defined `options` argument with `callback`
- Added tests cases to cover committed and aborted transactions

### CODE REVIEW

File: plugins/node/instrumentation-mongoose/test/mongoose.test.ts
Comment (ferrelucas) [inline plugins/node/instrumentation-mongoose/test/mongoose.test.ts:68]: ℹ️ Note this replacer function was only introduced here to avoid an error due to the Session object having cyclical reference

**Comment by ferrelucas** [comment]
Hi @blumamir, wondering if you could have a look at this PR when you have some time?

**Comment by ferrelucas** [comment]
@open-telemetry/javascript-approvers Hi there, wondering if somebody could have a look at this PR?

File: plugins/node/instrumentation-mongoose/src/utils.ts
Comment (blumamir) [inline plugins/node/instrumentation-mongoose/src/utils.ts:106]: The argument is an array but we only consume the first value which we expect to be an object. Can you explain the motivation for this?

Is `args[0]` the options object? in case why do we need to desturct it's values into an array? and what if it is being used with both options and a callback?

File: plugins/node/instrumentation-mongoose/src/utils.ts
Comment (blumamir) [inline plugins/node/instrumentation-mongoose/src/utils.ts:115]: are we overwriting a user callback which is now not invoked? should we store it's value before overwriting and invoke it if present from our response callback?

**Review by blumamir — COMMENTED** [review]
Sorry for the time it took to review and thank you for creating a fix for this issue 🙏

Not sure I fully understood the logic in this PR regarding adding new callback to the existing array.

File: plugins/node/instrumentation-mongoose/src/utils.ts
Comment (ferrelucas) [reply plugins/node/instrumentation-mongoose/src/utils.ts:106]: Hi @blumamir , thanks for taking the time to review!

Sorry, this was my mistake. I, for some reason, defined `args` as `...IArguments[]` and assumed it was necessary.
So, just for explaining the code:
Logging the the value of `args` when the method is called with both the `options` object and the `callback` function, we would get the following:
```
args: [ [Arguments] { '0': [Object], '1': [AsyncFunction (anonymous)] } ],
```
In this case, `args` would be an array that has only 1 element - the array of the method arguments: `options` and `callback`. Here is what loging the value of `args[0]` would result:
```
args0: [Arguments] {
    '0': { session: [ClientSession] },
    '1': [AsyncFunction (anonymous)]
  }
```
-----
But, yes, this is not necessary. 
I am pushing an update where we can make the update `args` is defined as `IArguments` and we can directly update `args`.

**Comment by ferrelucas** [comment]
@blumamir , I've update the PR to simplify the usage of `args`

Please let me know what you think
Thanks!

File: plugins/node/instrumentation-mongoose/src/utils.ts
Comment (blumamir) [reply plugins/node/instrumentation-mongoose/src/utils.ts]: Never mind, I was thinking that the `callback` is already part of the args and we can potentially extract it from it instead of sending it with the function. But now that I think about it again, I think it is more readable the way it is now 👍

**Review by blumamir — APPROVED** [review]
LGTM

Thanks again for the fix and sorry it took so long to review 🙏

**Comment by ferrelucas** [comment]
No worries!
Thanks for the review! 🙌

**Comment by blumamir** [comment]
@ferrelucas let me know if you need any help fixing the CI issues. once it is green we can merge and release this fix

**Comment by ferrelucas** [comment]
Hi @blumamir I've taken one of the CI unit tests as example `Run test-all-versions (20)`.
I've seen that the MongoDB setup is in the step called `Initialize containers`, under `Starting mongo service container`.

It is running:
```
/usr/bin/docker create --name fa2fac6564d8478ebd791968ddae63b9_mongo_11c61e --label 62ce9a --network github_network_60aacba5f9dc4f70be079739b8ce42a8 --network-alias mongo -p 27017:27017  -e GITHUB_ACTIONS=true -e CI=true mongo
```

I think we could add the `--replSet` parameter to the command to start the MongoDB instance, [then initialize the replica set](https://www.mongodb.com/resources/products/compatibilities/deploying-a-mongodb-cluster-with-docker).

I believe I don't have permissions to make these changes in the Github Actions
Would it be feasible to do it on your side?
Thanks!

**Comment by blumamir** [comment]
> I believe I don't have permissions to make these changes in the Github Actions
> Would it be feasible to do it on your side?

If I am not mistake, the mongo service is started [here](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/0cef9fa95d065489b7349bfa0150ff858ed59157/.github/workflows/unit-test.yml#L23) which you can edit, but I really think it would be easier to just use some other option for the test instead of `session`, if one of them is simpler.

**Comment by ferrelucas** [comment]
@blumamir thanks! I've updated the PR unit tests to use `wtimeout` instead of `session` in the options object

**Review by blumamir — APPROVED** [review]

**Comment by ferrelucas** [comment]
Hey, @blumamir thanks again for the review! 🙌

**Pull request merged**
